src/v/ssx/task_local_ptr.h

```diff
@@ -0,0 +1,161 @@
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "vassert.h"
+
+#include <seastar/core/reactor.hh>
+#include <seastar/core/task.hh>
+#include <seastar/core/thread.hh>
+
+#include <stdexcept>
+
+namespace ssx {
+
+/// Exception that indicates that the accessed
+/// task_local_ptr object is invalidated.
+class task_local_ptr_invalidated : public std::runtime_error {
+public:
+    task_local_ptr_invalidated()
+      : std::runtime_error("task_local_ptr is invalidated") {}
+};
+
+/// The value that gets invalidated after first scheduling point.
+///
+/// Any attempt to access the value after invalidation
+/// will trigger 'task_local_invalidated' exception.
+template<class T>
+class task_local {
+    static const seastar::task* get_current_task() {
+        auto task = seastar::engine().current_task();
+        if (task == nullptr) {
+            task = seastar::thread_impl::get()->waiting_task();
+        }
+        vassert(task != nullptr, "No current task");
+        return task;
+    }
+
+public:
+    explicit task_local(const T& val)
+      : _val(val)
+      , _task(get_current_task()) {}
+
+    explicit task_local(T&& val)
+      : _val(val)
+      , _task(get_current_task()) {}
+
+    task_local() = delete;
+    ~task_local() = default;
+    task_local(const task_local& other) = default;
+    task_local(task_local&&) noexcept = default;
+    task_local& operator=(const task_local&) = default;
+    task_local& operator=(task_local&&) noexcept = default;
+
+    bool operator==(const task_local& other) const {
+        throw_on_bad_access();
+        return _val == other._val;
+    }
+
+    bool operator==(const T& other) const {
+        throw_on_bad_access();
+        return _val == other;
+    }
+
+    const T& value() const {
+        check_time_slice();
+        throw_on_bad_access();
+        return _val.value();
+    }
+
+    T& value() {
+        check_time_slice();
+        throw_on_bad_access();
+        return _val.value();
+    }
+
+    bool has_value() const {
+        check_time_slice();
+        return _val.has_value();
+    }
+
+private:
+    /// Check if the time slice is the same. Invalidate
+    /// the pointer if its not.
+    void check_time_slice() const {
+        if (get_current_task() != _task) {
+            // Opt-out from all checks if we don't have a valid pointer to
+            // the current task.
+            _val = std::nullopt;
+        }
+    }
+
+    void throw_on_bad_access() const {
+        if (!_val.has_value()) {
+            throw task_local_ptr_invalidated();
+        }
+    }
+
+    mutable std::optional<T> _val;
+    // Non-owning ptr
+    const seastar::task* _task;
+};
+
+template<class T, class... Args>
+auto make_task_local(Args&&... args) {
+    return task_local<T>(T((std::forward<Args>(args))...));
```

Comment:
suggestion (non-blocking): this could use std::in_place_t  and delegate the construction at the inner std::optional

Reply:
will fix this in followup